framework/src/tinyalsa/tinyalsa.c

```diff
@@ -506,6 +504,18 @@ int pcm_writei(struct pcm *pcm, const void *data, unsigned int frame_count)
 				apb = (struct ap_buffer_s *)msg.u.pPtr;
 				memcpy(apb->samp, (char *)data + offset, nbytes);
 				apb->flags &= ~AUDIO_APB_OUTPUT_ENQUEUED;
+			} else if (msg.msgId == AUDIO_MSG_XRUN) {
+				/* Handle underrun */
+				/* If PCM_NORESTART flag is set, just return an ERROR. Let the application handle it */
+				if (pcm->flags & PCM_NORESTART) {
+					return -EPIPE;
+				}
+
+				/* Try to restart the codec and resend the data */
```

Comment:
preparing should be done by pcm_prepare.

Reply:
We'd better refer to tinyalsa's pcm.c
also AUDIOIOC_PREPARE should be defined in audio.h
We'd better do something in i2s & dma layer.
We will remove specific data from dma pool...

Reply:
We need to decide how to handle xrun scenario. 
How does the codec chip behave when an xrun happens? 
Do we need to restart it?
Or just do a sw reset of codec? 
Or nothing to be done?
In any case, I think we can use pcm_prepare instead of pcm_stop - pcm_start. And implement the required logic in codec side (since the behavior after xrun depends on each codec).

Reply:
No, application(or media f/w) should decide that recorver or not.
So we should provide pcm_prepare instead of pcm_stop & pcm_start.
pcm_prepare should send some message to device layer through the ioctl, and then device will do something when it get message from upper layer.
@btheosam and I have a plan that handle this in codec & i2s layer. we will design it.

Reply:
In other words, return -EPIPE is enough for pcm_writei(or call when it started like a original tinyalsa)
also pcm_prepare should call ioctl(pcm->fd, AUIDIOIC_PREPARE);
Application or media f/w will call as a below.
if (pcm->flags & PCM_MMAP) {
     ret = pcm_mmap_write();
} else {
    ret = pcm_write();
}

if (ret != 0) {
   switch(ret) {
      case XRUN : 
          ret = pcm_prepare(pcm);
}
if (pcm->flags & PCM_MMAP) {
     ret = pcm_mmap_write();
} else {
    ret = pcm_write();
}

Reply:
Ok. 
But in Tinyalsa implementation, pcm_prepare is being called directly inside pcm_writei and pcm_readi when XRUN happens. But mmap API's just return -EPIPE. Do you want to continue this type of implementation OR always return -EPIPE during XRUN?

Reply:
They check pcm->running before call pcm_prepare();
It means pcm_stopped due to some reasons, we can follow their policy.
In our case, if enqueue returns XRUN, we can just return -EPIPE. 
If you have more good idea, you can suggest but my opinion is that user should decice about recovery ...

Reply:
How about below implementation in tinyalsa?
In codec driver PREPARE ioctl can be handled in hw specific manner.

pcm_writei() {
...
	if(xrun) {
		if (pcm->flags & PCM_NORESTART) {
				return -EPIPE;
		} else {
			pcm_prepare();
			pending = pcm_frames_to_bytes(pcm, frame_count);
			offset = 0;
			continue;
		}
	}
}

pcm_readi() {
	...
	if (xrun) {
		if (!pcm->draining) {
			pcm_prepare();
		}
		continue;
	}
}

pcm_mmap_xxx_api() {
	...
	if (xrun) {
		return -EPIPE;
	}
}

Reply:
I totally agree with @Taejun-Kwon comment "No, application(or media f/w) should decide that recorver or not"